## Homework 6: Using LASSO Regression to Model Fat Sales Data
Hathaway Zhang <br>
104369396 <br>
Oct.4, 2018

In [33]:
import pandas as pd
import pandas
from sklearn.cross_validation import train_test_split
from sklearn.metrics import mean_squared_error
from sklearn.linear_model import LassoLarsCV
import matplotlib.pyplot as plt

In [3]:
dfRaw = pd.read_csv("finalmaster-ratios.csv")
dfRaw.head()

# Purchases  B01001001   B01001002  B01001003  B01001004  B01001005  \
0           22     206252  469.226965  31.432422  35.219052  33.628765   
1            7      61399  486.538869  22.899396  21.531295  27.036271   
2            3      73170  489.859232  28.905289  36.271696  28.235616   
3           94     251724  505.585483  32.054949  31.757004  28.102207   
4            0      37382  495.586111  25.413301  29.318924  26.162324   

   B01001006  B01001007  B01001008  B01001009    ...      B19001008  \
0  20.121017  12.610787   6.734480   6.225394    ...      49.409690   
1  16.808091  28.355511  18.192479  13.534422    ...      59.231680   
2  21.566216  12.218122   7.243406   7.380074    ...      63.996993   
3  18.651380  12.080692   7.035483   7.686991    ...      54.790900   
4  19.260607  12.893906   6.580707   7.062222    ...      58.883378   

   B19001009  B19001010  B19001011   B19001012   B19001013  B19001014  \
0  53.306757  42.318307  83.167229   89.249208  102.141470  52.872330   
1  50.093078  40.700626  92.612963  117.363344  113.344051  75.774243   
2  47.322923  42.505211  70.420610   90.033143   98.677692  54.703249   
3  48.681562  43.873381  84.717507  112.204444  127.137252  83.019904   
4  51.761414  47.310187  81.902582   93.793717  130.103014  71.982704   

   B19001015  B19001016  B19001017  
0  36.440765  23.446284  21.197485  
1  33.000508  33.169741  24.792689  
2  20.125056  11.890525  16.537397  
3  43.731067  38.851729  40.427349  
4  36.118530  31.603714  19.648989  

[5 rows x 190 columns]

In [26]:
# create a list of all the predictors you're going to feed into the LassoLarsCV model
allvariablenames = list(dfRaw.columns.values)
listofallpredictors = allvariablenames[8:]

In [27]:
#load predictors into dataframe
predictors = dfRaw[listofallpredictors]  
#load target into dataframe
target = dfRaw['# Purchases']   
target.head()

0    22
1     7
2     3
3    94
4     0
Name: # Purchases, dtype: int64

In [28]:
# split data into train and test sets, with 30% retained for test
pred_train, pred_test, tar_train, tar_test = train_test_split(predictors, target, test_size=.3, random_state=123) 
#  (1) pred_train, the predictors training set 
#  (2) pred_test, the predictors test test 
#  (3) tar_train, the target training set 
#  (4) tar_test, the target test set.
model = LassoLarsCV(cv=10, precompute=False).fit(pred_train, tar_train)

/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/least_angle.py:313: ConvergenceWarning: Regressors in active set degenerate. Dropping a regressor, after 5 iterations, i.e. alpha=1.496e+00, with an active set of 5 regressors, and the smallest cholesky pivot element being 5.771e-08. Reduce max_iter or increase eps parameters.
  ConvergenceWarning)
/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/least_angle.py:313: ConvergenceWarning: Regressors in active set degenerate. Dropping a regressor, after 8 iterations, i.e. alpha=1.098e+00, with an active set of 8 regressors, and the smallest cholesky pivot element being 5.771e-08. Reduce max_iter or increase eps parameters.
  ConvergenceWarning)
/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/least_angle.py:313: ConvergenceWarning: Regressors in active set degenerate. Dropping a regressor, after 10 iterations, i.e. alpha=7.329e-01, with an active set of 10 regressors, and the smallest cholesky pivot element b

/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/least_angle.py:313: ConvergenceWarning: Regressors in active set degenerate. Dropping a regressor, after 38 iterations, i.e. alpha=1.880e-01, with an active set of 36 regressors, and the smallest cholesky pivot element being 7.598e-08. Reduce max_iter or increase eps parameters.
  ConvergenceWarning)
/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/least_angle.py:339: ConvergenceWarning: Early stopping the lars path, as the residues are small and the current value of alpha is no longer well controlled. 43 iterations, alpha=1.742e-01, previous alpha=1.733e-01, with an active set of 38 regressors.
  ConvergenceWarning)
/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/least_angle.py:313: ConvergenceWarning: Regressors in active set degenerate. Dropping a regressor, after 4 iterations, i.e. alpha=1.372e+00, with an active set of 4 regressors, and the smallest cholesky pivot element being 3.799e-08. Reduce max

### QUESTION 1
In your own words, explain what the above lines of code are doing. Why am I doing it? Explain each line.

In [29]:
# build coefficent chart
# Question 1
# load predictors into predictors_model
predictors_model=pd.DataFrame(listofallpredictors)
# rename header of predictors_model as 'label'
predictors_model.columns = ['label']
# create a column named coeff and add coeff. to dataframe
predictors_model['coeff'] = model.coef_
# check all the index and row in predictors
for index, row in predictors_model.iterrows():
    # since any coefficients that are non-zero are significant, find all the coeff, greater than 0
    if row['coeff'] > 0:
        # print out the significant values
        print(row.values)

['B01001014' 0.8557934663949708]
['B01001036' 2.5053958893202446]
['B01001037' 0.8894158553356212]
['B01001038' 1.5315800095865324]
['B02001005' 0.41254033933206735]
['B13014026' 0.4800235646035068]
['B13014027' 0.6977472755369424]
['B13016001' 874929430.0943857]
['B19001017' 1.4834469570888447]


### Question 2
Interpret each variable intuitively. What Census variables most predict sales? What does that mean, practically? Here's what I'd put for the example above. "In areas where there are more females aged 30-34, we sell more Bobo Bars."

['B01001014' 0.8557908775529921] Males aged 40 to 44 Years. <br>
In areas where there are more males aged 40-44, we sell 0.8557908775529921 unit more Bobo Bars
<br>
<br>
['B01001036' 2.505392496591849] Females aged 30 to 34 Years. <br>
In areas where there are more females aged 30-34, we sell 2.505392496591849 unit more Bobo Bars
<br>
<br>
['B01001037' 0.8894214357013622] Females aged 35 to 39 Years. <br>
In areas where there are more females aged 35-39, we sell 0.8894214357013622] unit more Bobo Bars
<br>
<br>
['B01001038' 1.5315839680821497] Females aged 40 to 44 Years. <br>
In areas where there are more females aged 40-44, we sell 1.5315839680821497 unit more Bobo Bars
<br>
<br>
['B02001005' 0.4125408937426837] Asian Alone. <br>
In areas where there are asian alone, we sell 0.4125408937426837 unit more Bobo Bars
<br>
<br>
['B13014026' 0.4800240326923769] Women 15 to 50 Years Who Had a Birth in the Past 12 Months with Bachelor's Degree. <br>
In areas where there are more females aged 15-50 that had a birth in the past 12 months with Bachelor's degree, we sell 0.4800240326923769 unit more Bobo Bars
<br>
<br>
['B13014027' 0.6977454940063235] Women 15 to 50 Years Who Had a Birth in the Past 12 Months with Graduate or Professional Degree. <br>
In areas where there are more females aged 15-50 that had a birth in the past 12 months with graduate or professional degree, we sell 0.6977454940063235 unit more Bobo Bars
<br>
<br>
['B13016001' 874922971.7249781] Women 15 to 50 Years Who Had a Birth in the Past 12 Months. <br>
In areas where there are more females aged 15-50 that had a birth in the past 12 months, we sell 874922971.7249781 unit more Bobo Bars
<br>
<br>
['B19001017' 1.4834465563617387] Household with income 200,000 or More. <br>
In areas where there are more household with income 200,000 or more, we sell 1.4834465563617387 unit more Bobo Bars

### QUESTION 3
If I had to report only two census variables to my boss that most steeply predicted sales, what would those be?

Since the higher the value of the coefficient, the steeper the relationship between sales and that variable.<br>
I will report B13016001 and B01001036 to my boss.

### QUESTION 4
Run the code above, then do the same thing for your test sets (2) and (4) above. Compare the two outputs. Are the training and text set mean squared errors similar? What does that mean practically? Think back to your stats class, or Google it!

In [34]:
train_error_train = mean_squared_error(tar_train, model.predict(pred_train))
print ('training data MSE')
print(train_error_train)

training data MSE
22025.320770047212


In [36]:
train_error_test = mean_squared_error(tar_test, model.predict(pred_test))
print ('testing data MSE')
print(train_error_test)

testing data MSE
41549.13946199703


The mean squared error of training data set is different from the mean squared error of testing data set. Since the Mean Squared Error is a measure of how close a fitted line is to data points and the smaller the Mean Squared Error, the closer the fit is to the data, we could know that the training data set fits the model better than the testing data set does. 

(citation: https://www.vernier.com/til/1014/)

Repeat the same process for the test set. Compare the two R squareds. 

In [37]:
#r squared
rsquared_train=model.score(pred_train,tar_train)
print ('training data R-square')
print(rsquared_train)

training data R-square
0.24002799797602614


In [41]:
#r squared
rsquared_test=model.score(pred_test,tar_test)
print ('testing data R-square')
print(rsquared_test)

testing data R-square
0.17587095531914365


The training data is a better regression model as it has larger R-square value.

### QUESTION 5
If your boss asked, "How well does Census data, overall, predict sales?" What would you say? Why?

In general, Census data does a bad job predicting sales. In both MSE test and R-square test, the training date set does a better job fitting the model than the practical testing data set does. However, even within the training date set, the R-square value is still to small to make proper prediction. The R-square value (around 20%), tells the variance of its errors is 20% less than the variance of the dependent variable and the sd of its error is 11% less than the sd of the dependent variable. Therefore, the sd of the regression model’s errors is only about 1/10 the size of the sd of the errors that you would get with a constant-only model. Along with the fact that the MSE of both dataset is relatively large, we could conclude that the Census data couldn't predict the sales properly.

(citation: https://people.duke.edu/~rnau/rsquared.htm)

### QUESTION 6
Finally, let's see what our y-intercept is, so we can interpret what our baseline sales number looks like, all things considered: What is our baseline sales number? What does that mean, practically? Think back to what y-intercepts mean in regression models.

In [40]:
print("y interecept:")
print(model.intercept_)

y interecept:
22.194675187960804


The y-intercept is 22.194675187960804 in this cse. When x = 0, the corresponding y-value is the y-intercept. In this case, we could know that the baseline sales number is 22.194675187960804, which means when variables that truly predict unique amounts of sales equal to 0, the baseline sales are 22.194675187960804.